feat(monorepo): converts patternfly-react to a lerna monorepo

priley86 opened the pull request:
🕺 👯‍♂️ 👯 💃 🕺 

affects: `@patternfly/react-codemods, @patternfly/react-console, patternfly-react,
eslint-plugin-patternfly-react`

this change converts pf-react to use lerna/lerna-semantic-release and introduces monorepo folder
structure

<!--
Thanks for your interest in patternfly-react. We appreciate all issues filed and PRs submitted!

Please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What issue is being addressed here?) -->
**What**:
* adopts monorepo structure, similar to the one [here](https://github.com/atlassian/lerna-semantic-release)
* moves all `src` for current `patternfly-react` into `packages/core`
* updates build scripts for building sass/less/js in a monorepo
* adds @dmiller9911 's * [esm modules change](https://github.com/patternfly/patternfly-react/pull/291) to support esm and enable better code splitting downstream
* moves storybook global imports (`defaultTemplate`, `inlineTemplate`, `DOCUMENATION_URL` etc.) to be shared module imports (absolute) instead of relative
* moves travis release scripts into `.travis` and adopts similar scripts to @karelhala's  #252 
* adds @dmiller9911 's eslint changes which aggregates all existing lint rules and exposes them as a new eslint-plugin, consumable downstream and in the monorepo. 👏 👏 👏 
* preeminently adds @dmiller9911 's  codemod package, `@patternfly/react-codemods`, to show an example of how we  can use codemods for migrating imports downstream. This would work now if we choose to move `patternfly-react` over to `@patternfly/react-core` but will live this up to the community...
* ensures `@patternfly/react-console` is built with lerna

Some things I am still working on...shouldn't take me long... 🛁 
* ~updating README/CONTRIBUTING.md~
* ~adding in any changes that were merged this week. I was working in a sandbox [here](https://github.com/priley86/patternfly-react-monorepo) all week ;)~

cc: @jeff-phillips-18 @dmiller9911 @mturley @dgutride @karelhala 

Please let me know the final verdict on package namespace names here before we call this one done!

Closes: #201 #218  

<!-- Please provide a link to your fork's Storybook. See README notes on how to do this. -->
**Link to Storybook**:
https://priley86.github.io/patternfly-react

<!-- Are there any upstream issues or separate issues you need to reference? -->
**Additional issues**:


<!-- feel free to add additional comments -->
For anyone interested, I have compiled all of our Lerna/monorepo documentation in this doc and added some additional notes (step for step here).
https://docs.google.com/document/d/1w5_J1OpP4fQGLRP12U7v2DMi-yqBaOvPwrqHuUBI_3I/edit

AllenBW commented:
revenge of the pr 🍿 👀 👏

dmiller9911 commented on packages/console/package.json:
```diff
@@ -21,20 +25,24 @@
-    "build": "node ../../scripts/runScript.js build:console",
-    "prepublish": "yarn build"
+    "prebuild": "yarn build:scripts && yarn build:sass && yarn build:less",
+    "build:scripts": "babel src --out-dir dist/js --ignore .test.js,__mocks__",
```
this actually is probably not needed since the lerna build command should take care of it.

dmiller9911 commented on packages/console/package.json:
```diff
@@ -21,20 +25,24 @@
+    "prebuild": "yarn build:scripts && yarn build:sass && yarn build:less",
+    "build:scripts": "babel src --out-dir dist/js --ignore .test.js,__mocks__",
+    "build:less": "mkdir -p dist/less && cp -r less/* dist/less",
+    "build:sass": "mkdir -p dist/sass && cp -r sass/* dist/sass && node-sass --output-style compressed --include-path sass --include-path $npm_package_sassIncludes_patternfly --include-path $npm_package_sassIncludes_bootstrap --include-path $npm_package_sassIncludes_fontAwesome -o dist/css sass/console.scss"
```
These could probably live in the root too.  It would be nice to have it build similar to babel if possible.

priley86 replied:
i agree w/ this, but i'm not quite sure how to accomplish it since we have package specific [file].scss at the end... feel free to give this a go...but i don't think it's really a blocker.

dmiller9911 commented on packages/core/package.json:
```diff
@@ -0,0 +1,58 @@
+  },
+  "scripts": {
+    "prebuild": "yarn build:scripts && yarn build:sass && yarn build:less",
+    "build:scripts": "babel src --out-dir dist/js --ignore .test.js,__mocks__",
```
this is ran in the root.

dmiller9911 commented on storybook/webpack.config.js:
```diff
@@ -1,86 +1,61 @@
+const ROOT_DIR = '..';
+const PCKGS = `${ROOT_DIR}/packages`;
+
+const packages = readdirSync(path.resolve(__dirname, PCKGS))
```
this can actually be replaced with lernas package utils
```
const PackageUtilities = require('lerna/lib/PackageUtilities');
const Repository = require('lerna/lib/Repository');

const packages = PackageUtilities.getPackages(new Repository());
```
Each "package" includes the properties here: https://github.com/lerna/lerna/blob/master/core/package/index.js#L36

priley86 commented:
* made improvements mentioned by @dmiller9911 above
* added latest pf-react changes/story updates
* updated some documenation
* redeployed storybook
